include/onnxruntime/core/framework/data_types.h

```diff
@@ -497,7 +550,81 @@ class SparseTensorType : public SparseTensorTypeBase {
  private:
   SparseTensorType() {
     using namespace data_types_internal;
-    TensorElementTypeSetter<elemT>::SetSparseTensorElementType(mutable_type_proto());
+    TensorElementTypeSetter<elemT>::SetSparseTensorElementType(this->mutable_type_proto());
+  }
+};
+
+/// Common base-class for all optional types.
+class OptionalTypeBase : public DataTypeImpl {
+ public:
+  static MLDataType Type();
+
+  bool IsOptionalType() const override {
+    return true;
+  }
+
+  const OptionalTypeBase* AsOptionalType() const override {
+    return this;
+  }
+
+  bool IsCompatible(const ONNX_NAMESPACE::TypeProto& type_proto) const override;
+
+  size_t Size() const override {
+    // should never reach here.
+    ORT_NOT_IMPLEMENTED(__FUNCTION__, " is not implemented");
+  }
+
+  DeleteFunc GetDeleteFunc() const override {
+    // should never reach here.
+    ORT_NOT_IMPLEMENTED(__FUNCTION__, " is not implemented");
+  }
+
+  const ONNX_NAMESPACE::TypeProto* GetTypeProto() const override;
+
+  virtual MLDataType GetElementType() const {
+    // should never reach here.
+    ORT_NOT_IMPLEMENTED(__FUNCTION__, " is not implemented");
+  }
+
+  OptionalTypeBase(const OptionalTypeBase&) = delete;
+  OptionalTypeBase& operator=(const OptionalTypeBase&) = delete;
+
+ protected:
+  ONNX_NAMESPACE::TypeProto& mutable_type_proto();
+
+  OptionalTypeBase();
+  ~OptionalTypeBase() override;
+
+ private:
+  struct Impl;
+  Impl* impl_;
+};
+
+template <typename T, typename elemT>
+class OptionalType : public OptionalTypeBase {
+ public:
+  static_assert(data_types_internal::IsOptionalOrtType<T>::value,
+                "Requires one of the supported types: Tensor or TensorSeq");
+
+  static_assert(data_types_internal::IsTensorContainedType<elemT>::value,
+                "Requires one of the tensor fundamental types");
+
+  static MLDataType Type();
+
+  MLDataType GetElementType() const override {
+    if (std::is_same<T, Tensor>::value) {
+      return DataTypeImpl::GetTensorType<elemT>();
+    } else if (std::is_same<T, Tensor>::value) {
+      return DataTypeImpl::GetSequenceTensorType<elemT>();
+    } else {
+      // Will not reach here
+      ORT_ENFORCE(false, "Unsupported optional type");
+    }
```

Comment:
Can we use a static_assert here given the 'if/else if' are using compile time values?